pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +167,552 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+		if err != nil {
+			log.Error(err, logErrorPatchSecret, "secretName", secretName, "secretNamespace", externalSecret.Namespace)
+			syncCallsError.With(resourceLabels).Inc()
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// fetch existing secret (from the full cache)
+	// NOTE: we are using the `r.SecretClient` which we only use for managed secrets.
+	//       when `enableManagedSecretsCache` is true, this is a cached client that only sees our managed secrets,
+	//       otherwise it will be the normal controller-runtime client which may be cached or make direct API calls,
+	//       depending on if `enabledSecretCache` is true or false.
+	existingSecret := &v1.Secret{}
+	err = r.SecretClient.Get(ctx, client.ObjectKey{Name: secretName, Namespace: externalSecret.Namespace}, existingSecret)
 	if err != nil && !apierrors.IsNotFound(err) {
-		log.Error(err, errGetExistingSecret)
+		log.Error(err, logErrorGetSecret, "secretName", secretName, "secretNamespace", externalSecret.Namespace)
+		syncCallsError.With(resourceLabels).Inc()
 		return ctrl.Result{}, err
 	}
 
-	// refresh should be skipped if
-	// 1. resource generation hasn't changed
-	// 2. refresh interval is 0
-	// 3. if we're still within refresh-interval
-	if !shouldRefresh(externalSecret) && isSecretValid(existingSecret) {
-		refreshInt = (externalSecret.Spec.RefreshInterval.Duration - timeSinceLastRefresh) + 5*time.Second
-		log.V(1).Info("skipping refresh", "rv", getResourceVersion(externalSecret), "nr", refreshInt.Seconds())
-		return ctrl.Result{RequeueAfter: refreshInt}, nil
+	// ensure the full cache is up-to-date
+	// NOTE: this prevents race conditions between the partial and full cache.
+	//       we return an error so we get an exponential backoff if we end up looping,
+	//       for example, during high cluster load and frequent updates to the target secret by other controllers.
+	if secretPartial.UID != existingSecret.UID || secretPartial.ResourceVersion != existingSecret.ResourceVersion {
+		err = fmt.Errorf(errSecretCachesNotSynced, secretName)
+		log.Error(err, logErrorSecretCacheNotSynced, "secretName", secretName, "secretNamespace", externalSecret.Namespace)
+		syncCallsError.With(resourceLabels).Inc()
+		return ctrl.Result{}, err
 	}
-	if !shouldReconcile(externalSecret) {
-		log.V(1).Info("stopping reconciling", "rv", getResourceVersion(externalSecret))
-		return ctrl.Result{}, nil
+
+	// refresh will be skipped if ALL the following conditions are met:
+	// 1. refresh interval is not 0
+	// 2. resource generation of the ExternalSecret has not changed
+	// 3. the last refresh time of the ExternalSecret is within the refresh interval
+	// 4. the target secret is valid:
+	//     - it exists
+	//     - it has the correct "managed" label
+	//     - it has the correct "data-hash" annotation
+	if !shouldRefresh(externalSecret) && isSecretValid(existingSecret) {
+		log.V(1).Info("skipping refresh")
```

Comment:
This is not correct.

If we use `markAsDone`, we will potentially not honor the requested `refreshInterval` or may even just never reconcile at all:

1. Remember, a reconcile function is expected to be idempotent, as it is expected that we might run the reconcile when we do not need to (e.g. when the controller restarts as ES are reconciled).
2. `markAsDone()` makes the ExternalSecret show as "Ready" with "SecretSynced" message, and updates the `status.refreshTime` neither of which should be happening when we decide to skip the reconcile!

Reply:
Regardless, we need a way to show that reconcile run, otherwise, people might mistake the object to be skipped from the loop or not been dealt with for significantly longer than actually did.

I understand what you are saying, but we need some kind of update. Maybe add something like, lastRefreshed, just for indication? WDYT?

Reply:
It's fine for now. Let's tackle that in a different issue.